tools/migration-unified-annotation-versioning/migration.py

```diff
@@ -0,0 +1,529 @@
+import psycopg2
+import psycopg2.extras
+from psycopg2.extras import RealDictRow
+import math
+import logging
+import datetime
+from pathlib import Path
+import time
+from typing import Dict, Tuple, List, Optional, Callable, Set
+from rich.progress import track
+import msgspec
+import concurrent.futures
+import threading
+from functools import partial
+import heapq
+
+import fossildbapi_pb2 as proto
+import VolumeTracing_pb2 as Volume
+import SkeletonTracing_pb2 as Skeleton
+import Annotation_pb2 as AnnotationProto
+from utils import log_since, batch_range, format_duration, time_str
+from connections import connect_to_fossildb, connect_to_postgres, assert_grpc_success
+
+logger = logging.getLogger("migration-logs")
+checkpoint_logger = logging.getLogger("migration-checkpoints")
+
+
+LayerVersionMapping = Dict[str, Dict[int, int]] # tracing id to (old version to new version)
+MappingIdMap = Dict[str, str] # tracing id to editable mapping id
+
+
+class Migration:
+
+    def __init__(self, args):
+        logger.info(f"Initializing migration with args {args} ...")
+        self.args = args
+        self.src_stub = connect_to_fossildb(args.src, "source")
+        self.dst_stub = None
+        if not args.dry:
+            self.dst_stub = connect_to_fossildb(args.dst, "destination")
+        self.done_count = None
+        self.done_count_lock = threading.Lock()
+        self.failure_count = 0
+        self.failure_count_lock = threading.Lock()
+        self.total_count = None
+        self.before = 0
+
+    def run(self):
+        self.before = time.time()
+        annotations = self.read_annotation_list()
+        self.setup_checkpoint_logging()
+        self.done_count = 0
+        self.failure_count = 0
+        self.total_count = len(annotations)
+
+        with concurrent.futures.ThreadPoolExecutor(max_workers=self.args.num_threads) as executor:
+            executor.map(self.migrate_annotation, annotations)
+        if self.failure_count > 0:
```

Comment:
Good question! Since this is supposed to run unsupervised over the holidays, I thought it might be better to go on, so that we can possibly re-do the missing ones after applying fixes.

If *all* fail due to an immediate bug, we should immediately see that in the output.

What do you think?

Reply:
Ok. That sounds reasonable. Maybe still set the exit code to nonzero in case of failures?